operator-common/src/main/java/io/strimzi/operator/cluster/model/Ca.java

```diff
@@ -203,6 +206,11 @@ public Ca(Reconciliation reconciliation, CertManager certManager, PasswordGenera
         this.generateCa = generateCa;
         this.policy = policy == null ? CertificateExpirationPolicy.RENEW_CERTIFICATE : policy;
         this.renewalType = RenewalType.NOOP;
+        this.clock = Clock.systemUTC();
+    }
+
+    /* test */ protected void setClock(Clock clock) {
+        this.clock = clock;
```

Comment:
The main reason we need something like this (`Clock` or `Date`) in the `Ca` class is about checking the certificate expiration time which is always in UTC. The `Clock` class seems to work pretty well for testing purposes (and I guess it was the same reason why it was added in the `OpenSslCertManager` class). It easily allow to mock a specific time with the desired timezone. 
The Date supplier we are using in the KAO is mostly related to the maintenance time window and IIRC it was added for the same reason ... simplify testing of that feature. I think that having `Clock` there would be better as well.
I had a try using `Date` in my code instead of clock but the code would be doubled by passing a `Clock` to the `OpenSslCertManager` and then a corresponding `Date` instance (parsed from that Clock) to the `Ca`. For the purpose testing I don't think it's worth.

Reply:
A X509 cert's `notAfter` is, AFAIK, defined in UTC, so a supplier of "now" that's in UTC makes the most sense/is least confusing. `java.util.Date` has a timezone, so supplying "now" as a Date would create the need for adjusting to UTC, whereas `Clock.instant` is explictly in UTC already. (`java.util.Date` is basically a horrible API which should be avoided at all costs).

At a higher level we have the concept of maintenance times, which do honour timezones (`TZ` env var), but those are really just a means for enabling/disabling to code which then makes the real check. IIRC the use of quartz for the cron expression support requires using `java.util.Date` there. We could still pass in a Clock and convert it to a Date in order to evaluate the cron expr, but it would need testing. There's a case for doing that (consistency), but I don't think it should be a blocker for this PR, unless there's some nuance I'm not understanding.

Reply:
Ok, fair point. It is hard to argue to use something we already use if we already use both mechanisms :-). The Date supplier has the advantage that it is integrated through the whole reconciliation chain. If you plug into it, you can for example write complete tests going through `KafkaAsemblyOperator`. But I guess we do not need anything like that now, so I'm fine with it as it is (apart form removing the unused field).

PS: To make things more fun, `certNeedsRenewal` uses `System.currentTimeMillis()` ;-)

Reply:
I see two issues I will open to address later:

1. Using `Clock` supplier instead of the `Date` supplier which I think makes more sense (and it's consistent with what this PR uses).
2. Avoid to use the `System.currentTimeMillis()` in the `certNeedsRenewal` now that we have a `Clock` with this PR and we can do much better.

Reply:
Related to the usage we have with `Date`, we always need something in UTC AFAICS.
For the maintenance window we are comparing with the cron and the user is providing the expression as UTC.
We also use the Date supplier for timestamp in the CR condition and we are still formatting as UTC.
As Tom said, `Clock.instant` is explicitly in UTC already.

Reply:
Well, I understand why you want to use UTC for the certificate deletion. But you cannot just change it like that for the existing code. If it is not in UTC today, changing it to UTC would be a breaking change. So I think even if you change from Date to Clock, you need to maintain the same timezone behaviour for the places where date is used today.

Reply:
But today deletion is based on using UTC, because the expiration dates in the certs are in UTC and the `Instant.now()` we use today returns UTC. I am not changing any behaviour in the certificate expired removal, I am just allowing to inject the  'Clock` for testing purposes but during the normal usage of the operator, the code will behave as today.

Reply:
I know that this PR doesn't do it. I'm just saying that changing the Date supplier to Clock supplier if it doesn't use always the same timezone is not so easy for the maintenance time windows and other places where it is used today.

Reply:
Maybe we could chat offline about it and maybe I am misleading what you mean but I see the Date supplier used in two places ...

The first one is when writing the time in the CR conditions by using the `StatusUtils.iso8601` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/StatusUtils.java#L38) which takes the `Date` instance and write the time in UTC starting from it.

The second place is about the maintenance window in the `Util.isMaintenanceTimeWindowsSatisfied` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/Util.java#L725) where it sets the timezone on the cron expression as UTC and then uses the `isSatisfiedBy(date)`. Taking a look at the code inside this [method](https://github.com/quartz-scheduler/quartz/blob/master/quartz-core/src/main/java/org/quartz/CronExpression.java#L312) it seems that anyway when we pass our `Date` instance, it's getting the timezone of the cron expression (that we set to UTC) and compares dates in UTC anyway (even if our `Date` instance is referring to the local time).

So I was wondering ... because these are the only two places where we use the Date supplier and it's always handled in UTC, why not having "something" (Clock?) already taking the time in UTC by default?
Where do you see the Date supplier being used taking into account the local time and not always the UTC one?

Reply:
Well, your argument was that Date is not in UTC. All I'm saying is that whatever you change, the evaluation of the maintenance windows has to use the same timezoning as it uses today. Otherwise you can screw-up a lot of users. So if it uses UTC today, then its fine to use UTC with Clock instead of Date.

Reply:
Yeah the Date is not in UTC but when it comes to the maintenance window, everything is "moved" to be UTC. We set the cron expression to be UTC and the cron expression class itself set the UTC timezone on the Date instance (not UTC) we pass. At least this is how I read the `CroxExpression` class code above :-)

Reply:
Then it should be fine. This was my only concern. Otherwise I don't care that much if it uses Date or Clock.